ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java

```diff
@@ -72,15 +96,98 @@ protected Map<String, String> doScaleOut(String startUpArgs) {
-        FlinkOptimizer.class.getName(), jobPath, memory,
-        super.buildOptimizerStartupArgsString(resource));
+    long memory = jobManagerMemory + taskManagerMemory * resource.getThreadCount();
+    return String.format("%s/bin/flink run -m yarn-cluster -ytm %s -yjm %s %s -c %s %s -m %s %s",
+          getFlinkHome(), taskManagerMemory, jobManagerMemory, buildFlinkArgs(properties),
+          FlinkOptimizer.class.getName(), jobPath, memory,
+          super.buildOptimizerStartupArgsString(resource));
+  }
+
+  private Map<String, String> loadFlinkConfig() {
+    try {
+      return new Yaml().load(Files.newInputStream(Paths.get(getFlinkHome() + FLINK_CONFIG_PATH)));
+    } catch (IOException e) {
+      LOG.error("load flink conf yaml failed: {}", e.getMessage());
+      return Collections.emptyMap();
+    }
+  }
+
+  /**
+   * get jobManager and taskManager memory.
+   * An example of using Jobmanager memory parameters is as follows:
+   *    jobmanager.memory: 1024
+   *    flink-conf.jobmanager.memory.process.size: 1024M
+   *    flink-conf.yaml
+   * Prioritize from high to low.
+   *
+   */
+  protected long getMemorySizeValue(Map<String, String> properties, Map<String, String> flinkConfig,
+                                  String propertyKey, String finkConfigKey) {
+    return Optional.ofNullable(properties.get(propertyKey))
+        .map(Long::valueOf)
+        .orElseGet(() -> Optional.ofNullable(properties.get(FLINK_PARAMETER_PREFIX + finkConfigKey))
+            .map((v) -> parseMemorySize(v))
+            .orElseGet(() -> Optional.ofNullable(flinkConfig.get(finkConfigKey))
+                .map((v) -> parseMemorySize(v))
+                .orElse(0L)));
+  }
+
+  /**
+   * build user configured flink native parameters
+   */
+  protected String buildFlinkArgs(Map<String, String> properties) {
+    final Set<String> keysToFilter = new HashSet<>(Arrays.asList(
+        JOB_MANAGER_MEMORY_PROPERTY,
+        TASK_MANAGER_MEMORY_PROPERTY,
+        FLINK_PARAMETER_PREFIX + JOB_MANAGER_TOTAL_PROCESS_MEMORY,
+        FLINK_PARAMETER_PREFIX + TASK_MANAGER_TOTAL_PROCESS_MEMORY
+    ));
+
+    return properties.entrySet()
+        .stream().filter(entry -> !keysToFilter.contains(entry.getKey()))
+        .filter(entry -> entry.getKey().startsWith(FLINK_PARAMETER_PREFIX))
+        .map(entry -> "-yD " + entry.getKey().substring(FLINK_PARAMETER_PREFIX.length()) + "=" + entry.getValue())
+        .collect(Collectors.joining(" "));
+  }
+
+  /**
+   * memory conversion of units method, supporting m and g units
+   */
+  public long parseMemorySize(String memoryStr) {
+    if (memoryStr == null || memoryStr.isEmpty()) {
+      return 0;
+    }
+    memoryStr = memoryStr.toLowerCase().trim().replaceAll("\\s", "");
+    Matcher matcher = Pattern.compile("(\\d+)([mg])").matcher(memoryStr);
+    if (matcher.matches()) {
+      long size = Long.parseLong(matcher.group(1));
+      String unit = matcher.group(2);
+      switch (unit) {
```

Comment:
I'm a little worried about compatibility issues. You can refer to the analysis of flink conf [1], of course, this has satisfied the most commonly used unit (m, g).

[1]: https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/configuration/MemorySize.java#L352-L356

Reply:
> I'm a little worried about compatibility issues. You can refer to the parse of flink conf [1](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/configuration/MemorySize.java#L352-L356). Of course, this has satisfied the most commonly used unit (m, g).

At present, it is indeed only supported for `M` and `G`, because I believe there is no need to be compatible with units as small as `KB`. I believe that even testing such small resources is not enough, let alone production environments.

So I only support `M` and `G`.